### ISSUE TITLE
ruuvi_library_ringbuffer.c  ringbuffer_queue write lock checking  question

### ISSUE THREAD

**Issue by DG12** [ruuvi/ruuvi.libraries.c#3]
Should 
` if(!buffer->lock(buffer->writelock, false)) { return RUUVI_LIBRARY_ERROR_FATAL; }`
be BEFORE

 `buffer->head = ((buffer->head + 1) & buffer->index_mask);`

??
++++

In fact should the code be more like:

```
if(!buffer->lock(buffer->writelock, true)) { return RUUVI_LIBRARY_ERROR_CONCURRENCY; }
else buffer->lock(buffer->write lock)=true;

  memcpy((buffer->storage) + (buffer->head * buffer->block_size), 
         data, data_length);
  buffer->head = ((buffer->head + 1) & buffer->index_mask);

 buffer->lock(buffer->write lock)=false;
```

**Comment by ojousima** [comment]
Current code is intentional, it uses atomic compare-and-exchange function under the hood. This means that checking if the lock is free and locking the free lock is one operation which cannot be interrupted. 

If the lock is checked first and exchanged later it's possible that interrupt is triggered between check and set, and this results in error as the interrupt has written to buffer. 

There is another error in the code though, the lock should be taken first and only after the lock has been obtained the buffer fullness should be checked. Thanks for making me recheck the locking logic :)

**Comment by ojousima** [comment]
Fixed in #9e3a80a, thanks
